ansible/eda-server
# Add Sync Project Endpoint

## Description 

Adds an API endpoint to sync an existing project.

- refactor importing project files code
- add sync project endpoint `api/projects/<project_id>`

Currently syncs an existing project based on two factors:
1. If the content of an existing file has changed, the [code](https://github.com/ansible/eda-server/blob/0cea343e310d68d8651fd55c957daf8b7c74c524/src/eda_server/project.py#L488) updates the respective DB entry. 
2. If there is a new file created, the [code](https://github.com/ansible/eda-server/blob/0cea343e310d68d8651fd55c957daf8b7c74c524/src/eda_server/project.py#L492) imports the file into its respective table (rulebook, inventory, etc.)
3. TO BE IMPLEMENTED: If the existing file's name has changed but the content is the same.

## Testing

Testing this PR mainly involves importing a Github project and syncing new changes from the project. Feel free to use the following [test project](https://github.com/Dostonbek1/test-project-for-eda).
1. Fork [this Github project](https://github.com/Dostonbek1/test-project-for-eda)
2. Create an EDA project by calling POST on `http://localhost:9000/api/projects` with the following request body:
```
{
    "url": "<link to your forked project repo>",
    "name": "Test Project",
    "description": "Test project"
}
```
3. Check the DB tables (project, rulebook, inventory, etc.) to make sure the project files have been imported.
4. Push new changes to your forked repo; could be a new rulebook, extra_var, etc. files or change the contents of the existing files.
5. Send a POST request on `http://localhost:9000/api/projects/<project_id>`.
6. Verify the above request is successful.
7. Check DB and verify all changes from the project have been synced including the `git_hash` of the project.

Resolves: AAP-5704
